hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java

```diff
@@ -710,6 +703,8 @@ private void instantiateServices(boolean withNewSnapshot) throws IOException {
     prefixManager = new PrefixManagerImpl(metadataManager, isRatisEnabled);
     keyManager = new KeyManagerImpl(this, scmClient, configuration,
         omStorage.getOmId());
```

Comment:
> But the keyManager, volumeManager fields etc are also changed by instantiateServices(), and they are not defined to be "volatile". That seems like a bug to me, but perhaps it was intentially done for performance reasons.

According to [Oracle's Java doc](https://docs.oracle.com/javase/tutorial/essential/concurrency/atomic.html), reference variables reads and writes are atomic even without the `volatile` keyword:

> - Reads and writes are atomic for reference variables and for most primitive variables (all types except long and double).
> - Reads and writes are atomic for all variables declared volatile (including long and double variables).

`volumeManager` and `keyManager` are reference variables. So it should be fine, in theory. CMIIW.

Reply:
It is true that the writes are atomic.  The problem is that Java makes no guarantee that the writes will be visible to other threads, if they aren't properly "synchronized", either with the volatile keyword or by using synchronized blocks.

Here is a relatively short blog post explaining the issue:
https://www.logicbig.com/tutorials/core-java-tutorial/java-multi-threading/visibility-volatile.html

The best explanation I've found is in my favorite java book, "Java Concurrency In Practice", chapter three, but it is a bit long:
https://leon-wtf.github.io/doc/java-concurrency-in-practice.pdf#%5B%7B%22num%22%3A63%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C-6%2C466%2C1%5D

Reply:
Thanks @GeorgeJahad . Good read.

I see how stale read could potentially become a problem now. e.g. after another `instantiateServices()` call while other threads are still actively using `metadataManager`, `multiTenantManager`, `volumeManager`, etc. This might cause legitimate issue in worker threads like `OpenKeyCleanupService`, `KeyDeletingService`. Thanks for raising awareness of this potential issue.

IIRC, making those references `volatile` shouldn't impose much performance impact on variable reads (at least on x86 because of [TSO](https://docs.oracle.com/cd/E23824_01/html/819-3196/hwovr-15.html#:~:text=store%20ordering%20(PSO).-,Total%20Store%20Ordering,-(TSO))). Let's open a separate jira on master branch to discuss this with others as well.